Add documentation for configurations (#106)
Leaf-Agriculture/docs: docs/configurations_endpoints.md

## Endpoints

### Get Api Owner's Configuration

&nbsp<span class="badge badge--success">GET</span> `/configs`

Gets the configuration of the Api Owner.

#### Response
A JSON containing the configuration of the Api Owner.

<Tabs
  defaultValue="sh"
  values={[
    { label: 'cURL', value: 'sh', },
    { label: 'Python', value: 'py', },
    { label: 'JavaScript', value: 'js', },
    { label: 'JSON Response', value: 'res', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'

  const endpoint ='https://api.withleaf.io/services/config/api/configs'
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  axios.get(endpoint, { headers })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'

  endpoint = 'https://api.withleaf.io/services/config/api/configs'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  response = requests.get(endpoint, headers=headers)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN

>   curl -X GET \
>       -H 'Authorization: Bearer ${TOKEN}' \
>       'https://api.withleaf.io/service/configs/api/configs'

Reviewer: Use double quote to allow the variables to be processed
```suggestion
  curl -X GET \
      -H "Authorization: Bearer ${TOKEN}" \
      "https://api.withleaf.io/service/configs/api/configs"
```

#### Response
A JSON containing the configuration of the Leaf User.

<Tabs
  defaultValue="sh"
  values={[
    { label: 'cURL', value: 'sh', },
    { label: 'Python', value: 'py', },
    { label: 'JavaScript', value: 'js', },
    { label: 'JSON Response', value: 'res', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'
  const LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  const endpoint = `https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}`
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  axios.get(endpoint, { headers })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'
  LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  endpoint = f'https://api.withleaf.io/services/config/api/configs/{LEAF_USER_ID}'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  response = requests.get(endpoint, headers=headers)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN
  LEAF_USER_ID=00000000-0000-0000-0000-000000000000
  

>   curl -X GET \
>       -H 'Authorization: Bearer ${YOUR_TOKEN}' \
>       'https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}'

Reviewer: Incorrect variable name and single quote won't replace the variable
```suggestion
  curl -X GET \
      -H "Authorization: Bearer ${TOKEN}" \
      "https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}"
```

  ```

  </TabItem>
  <TabItem value="res">

  ```json
  {
    "apiOwnerUsername": "api-owner",
    "leafUserId": "00000000-0000-0000-0000-000000000000",
    "operationsSummaryCreation": true,
    "operationImageCreation": true,
    "fieldsAutoSync": true
  }
  ```

  </TabItem>
</Tabs>

### Create Leaf User's Configuration

&nbsp<span class="badge badge--warning">POST</span> `/configs/{leafUserId}`

> Creates the Configuration for the Leaf User `leafUserId`. A resquest body must be provided
> containing the configurations to be set. All entries are optional, if missing the configuration will be inherited from the Api Owner's Configuration.

Reviewer: ```suggestion
Creates the Configuration for the Leaf User `leafUserId`. A request body must be provided
containing the configurations to be set. All entries are optional, any missing configuration will be inherited from the Api Owner's Configuration.
```

  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'
  const LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  const endpoint = `https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}`
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  const data = {
    "operationImageCreation": true,
    "operationsSummaryCreation": true,
    "fieldsAutoSync": true
  }

  axios.post(endpoint, { headers, data })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'
  LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  endpoint = f'https://api.withleaf.io/services/config/api/configs/{LEAF_USER_ID}'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  data = {
    'operationImageCreation': true,
    'operationsSummaryCreation': true,
    'fieldsAutoSync': true
  }

  response = requests.post(endpoint, headers=headers, json=data)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN
  LEAF_USER_ID=00000000-0000-0000-0000-000000000000
  

>   curl -X POST \
>       -H 'Authorization: Bearer ${TOKEN}' \
>       -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
>       'https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}'

Reviewer: Use double quote to allow the variables to be processed
```suggestion
  curl -X POST \
      -H "Authorization: Bearer ${TOKEN}" \
      -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
      "https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}"
```

  ```

  </TabItem>
  <TabItem value="res">

  ```json
  {
    "apiOwnerUsername": "api-owner",
    "leafUserId": "00000000-0000-0000-0000-000000000000",
    "operationsSummaryCreation": true,
    "operationImageCreation": true,
    "fieldsAutoSync": true
  }
  ```

  </TabItem>
</Tabs>

### Update Api Owner's Configuration

&nbsp<span class="badge badge--success">PUT</span> `/configs`

> Updates the Configuration for the Api Owner. A resquest body must be provided
> containing the configurations to be set. All entries are required.

Reviewer: ```suggestion
Updates the Configuration for the Api Owner. A request body must be provided
containing the configurations to be set. All entries are required.
```

<Tabs
  defaultValue="sh"
  values={[
    { label: 'cURL', value: 'sh', },
    { label: 'Python', value: 'py', },
    { label: 'JavaScript', value: 'js', },
    { label: 'JSON Response', value: 'res', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'

  const endpoint ='https://api.withleaf.io/services/config/api/configs'
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  const data = {
    "operationImageCreation": true,
    "operationsSummaryCreation": true,
    "fieldsAutoSync": true
  }

  axios.put(endpoint, { headers, data })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'

  endpoint = 'https://api.withleaf.io/services/config/api/configs'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  data = {
    'operationImageCreation': true,
    'operationsSummaryCreation': true,
    'fieldsAutoSync': true
  }

  response = requests.put(endpoint, headers=headers, json=data)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN = 'YOUR_TOKEN'

>   curl -X PUT \
>       -H 'Authorization: Bearer ${TOKEN}' \
>       -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
>       'https://api.withleaf.io/services/config/api/configs'

Reviewer: Use double quote to allow the variables to be processed
```suggestion
  curl -X PUT \
      -H "Authorization: Bearer ${TOKEN}" \
      -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
      'https://api.withleaf.io/services/config/api/configs'
```

  ```

  </TabItem>
  <TabItem value="res">

  ```json
  {
    "apiOwnerUsername": "api-owner",
    "leafUserId": "",
    "operationsSummaryCreation": true,
    "operationImageCreation": true,
    "fieldsAutoSync": true
  }
  ```

  </TabItem>
</Tabs>

### Update Leaf User's Configuration

&nbsp<span class="badge badge--success">PUT</span> `/configs/{leafUserId}`

> Updates the Configuration for the Leaf User `leafUserId`. A resquest body must be provided
> containing the configurations to be set. All entries are optional, if missing the configuration will be inherited from the Api Owner's Configuration.

Reviewer: ```suggestion
Updates the Configuration for the Leaf User `leafUserId`. A request body must be provided
containing the configurations to be set. All entries are optional, any missing configuration will be inherited from the Api Owner's Configuration.
```

  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'
  const LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  const endpoint = `https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}`
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  const data = {
    "operationImageCreation": true,
    "operationsSummaryCreation": true,
    "fieldsAutoSync": true
  }

  axios.put(endpoint, { headers, data })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'
  LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  endpoint = f'https://api.withleaf.io/services/config/api/configs/{LEAF_USER_ID}'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  data = {
    'operationImageCreation': true,
    'operationsSummaryCreation': true,
    'fieldsAutoSync': true
  }

  response = requests.put(endpoint, headers=headers, json=data)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN
  LEAF_USER_ID=00000000-0000-0000-0000-000000000000

>   curl -X PUT \
>       -H 'Authorization: Bearer ${TOKEN}' \
>       -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
>       'https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}'

Reviewer: Use double quote to allow the variables to be processed
```suggestion
  curl -X PUT \
      -H "Authorization: Bearer ${TOKEN}" \
      -d '{ "operationImageCreation": true, "operationsSummaryCreation": true, "fieldsAutoSync": true }' \
      "https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}"
```

  ```

  </TabItem>
  <TabItem value="res">

  ```json
  {
    "apiOwnerUsername": "api-owner",
    "leafUserId": "00000000-0000-0000-0000-000000000000",
    "operationsSummaryCreation": true,
    "operationImageCreation": true,
    "fieldsAutoSync": true
  }
  ```

  </TabItem>
</Tabs>

> ### Delete Leaf User's Configuraiton

Reviewer: ```suggestion
### Delete Leaf User's Configuration
```

<Tabs
  defaultValue="sh"
  values={[
    { label: 'cURL', value: 'sh', },
    { label: 'Python', value: 'py', },
    { label: 'JavaScript', value: 'js', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'
  const LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  const endpoint = `https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}`
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  axios.delete(endpoint, { headers })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'
  LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  endpoint = f'https://api.withleaf.io/services/config/api/configs/{LEAF_USER_ID}'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  response = requests.delete(endpoint, headers=headers)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN
  LEAF_USER_ID=00000000-0000-0000-0000-000000000000

>   curl -X DELETE \
>       -H 'Authorization: Bearer ${TOKEN}' \
>       'https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}'

Reviewer: Use double quote to allow the variables to be processed
```suggestion
  curl -X DELETE \
      -H "Authorization: Bearer ${TOKEN}" \
      "https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}"
```

}
```

#### Response
A JSON containing the configuration of the Api Owner.

<Tabs
  defaultValue="sh"
  values={[
    { label: 'cURL', value: 'sh', },
    { label: 'Python', value: 'py', },
    { label: 'JavaScript', value: 'js', },
    { label: 'JSON Response', value: 'res', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'

  const endpoint ='https://api.withleaf.io/services/config/api/configs'
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  const data = {
    "operationsImageCreation": true,
    "fieldsAutoSync": true
  }

  axios.patch(endpoint, { headers, data })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'

  endpoint = 'https://api.withleaf.io/services/config/api/configs'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  data = {
    'operationsImageCreation': true,
    'fieldsAutoSync': true
  }

  response = requests.patch(endpoint, headers=headers, json=data)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN = 'YOUR_TOKEN'

  curl -X PATCH \

>       -H 'Authorization: Bearer ${TOKEN}' \

Reviewer: Variables inside a single quote won't be replaced 😉 
```suggestion
      -H 'Authorization: Bearer ${TOKEN}' \
```

Author: Fixed on bc35a888d7494b55fad60f05dc5c81adcc62c716

    { label: 'cURL', value: 'sh', },
    { label: 'JSON Response', value: 'res', },
  ]
}>
  <TabItem value="js">

  ```js
  const axios = require('axios')
  const TOKEN = 'YOUR_TOKEN'
  const LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  const endpoint = `https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}`
  const headers = { 'Authorization': `Bearer ${TOKEN}` }

  const data = {
    "operationsImageCreation": true,
    "fieldsAutoSync": true
  }

  axios.patch(endpoint, { headers, data })
      .then(res => console.log(res.data))
      .catch(console.error)
  ```

  </TabItem>
  <TabItem value="py">

  ```py
  import requests

  TOKEN = 'YOUR_TOKEN'
  LEAF_USER_ID = '00000000-0000-0000-0000-000000000000'

  endpoint = f'https://api.withleaf.io/services/config/api/configs/{LEAF_USER_ID}'
  headers = {'Authorization': f'Bearer {TOKEN}'}

  data = {
    'operationsImageCreation': true,
    'fieldsAutoSync': true
  }

  response = requests.patch(endpoint, headers=headers, json=data)
  print(response.json())
  ```

  </TabItem>
  <TabItem value="sh">

  ```shell
  TOKEN=YOUR_TOKEN
  LEAF_USER_ID=00000000-0000-0000-0000-000000000000

  curl -X PATCH \

>       -H 'Authorization: Bearer ${TOKEN}' \

Reviewer: Variables inside a single quote won't be replaced 😉 
```suggestion
      -H "Authorization: Bearer ${TOKEN}" \
```

Author: Fixed on bc35a888d7494b55fad60f05dc5c81adcc62c716

      -d '{ "operationsImageCreation": true, "fieldsAutoSync": true }' \

>       'https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}'

Reviewer: Variables inside a single quote won't be replaced 😉 
```suggestion
      "https://api.withleaf.io/services/config/api/configs/${LEAF_USER_ID}"
```

Author: I was replicating a very nasty behavior that shouldn't be common in the docs, but it is, and that's not actually running those code snippets. I ran each of the snippets individually now, they all run successfully after some changes. We have more wrong snippets across the docs, this should be reviewed. bc35a888d7494b55fad60f05dc5c81adcc62c716

Reviewer: Ouch, nice catch!

[1]: https://github.com/Leaf-Agriculture/Leaf-quickstart-Postman-collection
[2]: #get-api-owner-config
[3]: #get-leaf-user-config
[4]: #create-leaf-user-config
[5]: #override-api-owner-config
[6]: #override-leaf-user-config
[7]: #update-api-owner
[8]: #update-leaf-user-config
[9]: #delete-leaf-user-config

## About
Here we list all the available endpoints from Configuration API. For easily
calling them, we recommend using [Leaf's Postman collection][1].

All HTTP methods should be prepended by this service's endpoint:

```
https://api.withleaf.io/services/config/api
```

This service has the following endpoints available:

Description | Endpoints
--- | ---
[Get Api Owner's Configuration][2] | <span class="badge badge--success">GET</span> `/configs`
[Get Leaf User's Configuration][3] | <span class="badge badge--success">GET</span> `/configs/{leafUserId}`
[Create Leaf User's Configuration][4] | <span class="badge badge--warning">POST</span> `/configs/{leafUserId}`
[Override Api Owner's Configuration][5] | <span class="badge badge--success">PUT</span> `/configs`
[Override Leaf User's Configuration][6] | <span class="badge badge--success">PUT</span> `/configs/{leafUserId}`

> [Update Leaf User's Configuraiton][7] | <span class="badge badge--warning">PATCH</span> `/configs/{leafUserId}`

Reviewer: ```suggestion
[Update Api Owner's Configuraiton][7] | <span class="badge badge--warning">PATCH</span> `/configs/{leafUserId}`
```
